torch_xla/csrc/xla_graph_executor.cpp

```diff
@@ -636,18 +626,76 @@ XLAGraphExecutor::ExecuteComputationWithBarrier(
                "(XLATensor).";
         dataptr = torch_xla::TensorToXlaData(ivalue.toTensor(), device);
       }
-
       ++idx;
       arguments.push_back(dataptr);
     }
   }
 
+  // Mimic the `PrepareOutputShardingPropagation` function at
+  // xla_sharding_util.cpp. One subtle difference is that in this path, we don't
+  // have explicit `tensors`. However, we already have `output_shapes` and
+  // `device` which were what the `tensors` were used for in the original
+  // `PrepareOutputShardingPropagation` function.
+  std::vector<XLATensor::ShardingSpecPtr> sharding_specs(placeholders.size());
+  if (ShardingUtil::UseVirtualDevice()) {
+    auto computation_proto =
+        cachedComputation->computation->computation().proto();
+    std::vector<xla::OpSharding> output_shardings;
+    if (computation_proto.has_spmd_output_sharding()) {
```

Comment:
can we put everything in a helper function instead? This function is getting way too long and it is hard to read.

Reply:
Updated